File: golang/1605.find-valid-matrix-given-row-and-column-sums.go
func restoreMatrix(rowSum []int, colSum []int) [][]int {
	m, n := len(rowSum), len(colSum)

	mat := make([][]int, m)
	for i := range mat {
		mat[i] = make([]int, n)
	}

	// O(max(m, n))
	i, j := 0, 0
	for i < m && j < n {
		mat[i][j] = min(rowSum[i], colSum[j])
		rowSum[i] -= mat[i][j]
		colSum[j] -= mat[i][j]

		if rowSum[i] == 0 {
			i++
		}

		if colSum[j] == 0 {
			j++
		}
	}

	return mat
}

File: python/1605.find-valid-matrix-given-row-and-column-sums.py
class Solution:
    def restoreMatrix(self, rowSum: list[int], colSum: list[int]) -> list[list[int]]:
        m, n = len(rowSum), len(colSum)
        matrix = [[0] * n for _ in range(m)]
        i, j = 0, 0

        while i < m and j < n:
            matrix[i][j] = min(rowSum[i], colSum[j])
            if rowSum[i] == colSum[j]:
                i += 1
                j += 1
            elif rowSum[i] > colSum[j]:
                rowSum[i] -= colSum[j]
                j += 1
            else:
                colSum[j] -= rowSum[i]
                i += 1

        return matrix

